feat: function tools in OpenAI Responses

bbrowning opened the pull request:
# What does this PR do?

This is a combination of what was previously 3 separate PRs - #2069, #2075, and #2083. It turns out all 3 of those are needed to land a working function calling Responses implementation. The web search builtin tool was already working, but this wires in support for custom function calling.

I ended up combining all three into one PR because they all had lots of merge conflicts, both with each other but also with #1806 that just landed. And, because landing any of them individually would have only left a partially working implementation merged.

The new things added here are:
* Storing of input items from previous responses and restoring of those input items when adding previous responses to the conversation state
* Handling of multiple input item messages roles, not just "user" messages.
* Support for custom tools passed into the Responses API to enable function calling outside of just the builtin websearch tool.

Closes #2074
Closes #2080

## Test Plan

### Unit Tests

Several new unit tests were added, and they all pass. Ran via:

```
python -m pytest -s -v tests/unit/providers/agents/meta_reference/test_openai_responses.py
```

### Responses API Verification Tests

I ran our verification run.yaml against multiple providers to ensure we were getting a decent pass rate. Specifically, I ensured the new custom tool verification test passed across multiple providers and that the multi-turn examples passed across at least some of the providers (some providers struggle with the multi-turn workflows still).

Running the stack setup for verification testing:

```
llama stack run --image-type venv tests/verifications/openai-api-verification-run.yaml
```

Together, passing 100% as an example:

```
pytest -s -v 'tests/verifications/openai_api/test_responses.py' --provider=together-llama-stack
```

## Documentation

We will need to start documenting the OpenAI APIs, but for now the Responses stuff is still rapidly evolving so delaying that.

ashwinb commented on llama_stack/apis/agents/openai_responses.py:
```diff
@@ -17,6 +17,28 @@ class OpenAIResponseError(BaseModel):
+    image_url: str | None = None
+
+
+# TODO: handle file content types
```
nit: also audio?

bbrowning replied:
To this general point, one of the next things I'd like to do is stub in the entire Responses API surface even if it's not all implemented yet.

However, on this specific point, the Responses API does not yet support audio. That's only in Chat Completion API for now, although I expect it will get added to Responses at some point. But, for now, only image, text ,and file content types are supported in Responses.

derekhiggins commented on tests/unit/providers/agents/meta_reference/fixtures/__init__.py:
```diff
@@ -0,0 +1,74 @@
+    with open(fixture_path) as f:
+        data = yaml.safe_load(f)
+
+    choices = []
```
Makes sense, (This was part of my PR that ben pulled in), 
@bbrowning  here is a commit that applies to your branch, https://goodsquishy.com/upload/65ef6429db7513d5b48a

feel free to squash it into the original unit test commit if you want it

bbrowning replied:
Ahh, thanks for that! I got caught up in some other things yesterday so didn't get a chance to address this. Applied your patch on top.

ashwinb commented on llama_stack/providers/inline/agents/meta_reference/openai_responses.py:
```diff
@@ -50,15 +56,46 @@
+        for input_message in input:
+            if isinstance(input_message.content, list):
+                content = []
+                for input_message_content in input_message.content:
```
this is similar in spirit to my earlier comments. is this `ContentText` to `ContentPartTextPaam` conversion any useful? the shape of these dicts is exactly the same (and has to be) right? these types are an artifact of the Stainless SDK generation (openai also generates their SDK via Stainless)

I want to see if we can reduce complexity here if not really warranted.

bbrowning replied:
The shape is actually not exactly the same here - what this is doing is converting the OpenAI Responses API input field into a Chat Completions messages field. This is because we convert Responses requests into Chat Completion requests for the actual inference calls.

So, we're iterating through the possible types of valid Response input and converting them to their Chat Completion equivalent.

bbrowning replied:
I normalized the names of these methods converting Responses API inputs -> Chat Completion messages and Chat Completion choices -> Responses API outputs. And, added a brief comment for each describing what it's converting to/from.

And, specifically for the chat content here, refactored the core part of that into a function just for converting these content parts with a note that they look similar, but are not identical.

ashwinb commented on llama_stack/apis/agents/openai_responses.py:
```diff
@@ -89,45 +130,59 @@ class OpenAIResponseObjectStreamResponseCompleted(BaseModel):
+    object: Literal["list"] = "list"
+
+
+class OpenAIResponsePreviousResponseWithInputItems(BaseModel):
```
this doesn't need to be in the API file and can stay next to the implementation only

ashwinb commented on llama_stack/providers/inline/agents/meta_reference/openai_responses.py:
```diff
@@ -90,19 +138,44 @@ def __init__(
+            new_input_items.extend(previous_response_with_input.response.output)
+
+            # new input items from the current request
+            if isinstance(input, str):
```
there could be other inputs -- namely the list? why are they not added?

bbrowning replied:
All this is doing is ensuring if the input is a string, it gets converted to a list of size 1. When the input is already a list, we just use it as-is. However, re-reading this I get your point, and will adjust this so that if it's a string we `new_input_items.append(...)` else we `new_input_items.extend(...)`. That will make what this is doing more clear I believe.

bbrowning replied:
I adjusted these lines to something that makes it more clear what's happening here - 
```
            if isinstance(input, str):
                new_input_items.append(OpenAIResponseMessage(content=input, role="user"))
            else:
                new_input_items.extend(input)
```

That is cleaner that what I had previously, and makes it much more clear that we're handling both sides of that `if` statement.

ashwinb commented on llama_stack/providers/inline/agents/meta_reference/openai_responses.py:
```diff
@@ -181,10 +231,28 @@ async def create_openai_response(
-            )
+        # TODO: should we check more than choices[0] here?
+        if chat_response.choices[0].message.tool_calls and tools:
+            # TODO: Should we support a mix of custom and builtin tools?
```
seems OK for now to not support this

bbrowning replied:
I agree with regards to mixing custom / builtin tools. I adjusted this comment from a TODO to `# Assume if the first tool is a function, all tools are functions` just so it's clear in case we find a need to support a mix of built-in and custom tools within a single request as users try things out.

Also, it reminded me of the `TODO: should we check more than choices[0] here?` right above there, which I fixed by just iterating over the choices instead of taking `choices[0]`.

ashwinb commented on tests/unit/providers/agents/meta_reference/fixtures/__init__.py:
```diff
@@ -0,0 +1,31 @@
+    Returns:
+        OpenAIChatCompletion object
+    """
+    fixtures_dir = os.path.dirname(os.path.abspath(__file__))
```
this could go as a const at the top of the file

derekhiggins replied:
yup, Makes sense

ashwinb commented on tests/unit/providers/agents/meta_reference/fixtures/__init__.py:
```diff
@@ -0,0 +1,31 @@
+
+
+def load_chat_completion_fixture(filename: str) -> OpenAIChatCompletion:
+    """
```
this comment block is quite unnecessary. this is a short clear function.

derekhiggins commented:
Functionality looks good, tested loading previous state and web_search tool calls

bbrowning commented:
Thanks for all the feedback! I believe I've addressed everything raised, but feel free to let me know if there are more changes or concerns. Sorry for the rebase in the middle of addressing feedback, but there was a conflict with other changes in the agents implementation that had to be resolved.

bbrowning commented:
With the changes as they stand now, I'm able to get basic workflows working with OpenAI Codex pointed to Llama Stack and Llama 4 Maverick models hosted on one of our supported providers. OpenAI Codex here is used as a way to explore and find bugs in our Responses API implementation, since it's speaking to Llama Stack over that API. A basic demo of that at https://asciinema.org/a/718486 .

ashwinb commented on llama_stack/providers/inline/agents/meta_reference/openai_responses.py:
```diff
@@ -50,31 +60,110 @@
-        if isinstance(output_message, OpenAIResponseOutputMessage):
-            messages.append(OpenAIAssistantMessageParam(content=output_message.content[0].text))
+    if isinstance(input, list):
+        for input_message in input:
```
nit: I don't think this variable should be called `input_message` -- maybe `input_item` is better

bbrowning replied:
Agreed - push a small update to rename this.

ashwinb commented on llama_stack/providers/inline/agents/meta_reference/openai_responses.py:
```diff
@@ -163,7 +256,26 @@ async def create_openai_response(
+                                chat_response_tool_calls[tool_call.index] = OpenAIChatCompletionToolCall(
+                                    **tool_call.model_dump()
+                                )
+                            chat_response_tool_calls[tool_call.index].function.arguments = (
```
nit: I think this can be written a bit more compactly perhaps by assigning an intermediate variable

bbrowning replied:
Yes, you're right. I added a code comment here about what this does and an intermediate variable as tool call args are accumulated.

ashwinb reviewed (approved):
lgtm thank you for the iteration!
